File: Leetcode/Python3/393. UTF-8 Validation/solution.py
# https://leetcode.com/problems/utf-8-validation/


class Solution:
    def validUtf8(self, data: List[int]) -> bool:
        def check(idx: int, count: int) -> bool:
            for i in range(1, count + 1):
                if idx >= len(data) or data[idx + i] >> 6 != 0b10:
                    return False
            return True  # Vaild UTF-8

        index = 0
        flag = True
        while index < len(data) and flag:
            bits = data[index]
            if bits >> 7 == 0b0:
                index += 1
            elif bits >> 5 == 0b110:
                flag = check(index, 1)
                index += 2
            elif bits >> 4 == 0b1110:
                flag = check(index, 2)
                index += 3
            elif bits >> 3 == 0b11110:
                flag = check(index, 3)
                index += 4
            else:   # If not vaild UTF-8 exist
                return False
        return flag

File: Leetcode/C++/393. UTF-8 Validation/solution.cpp
// https://leetcode.com/problems/utf-8-validation/

#include <vector>

using namespace std;

class Solution
{
public:
    bool validUtf8(vector<int> &data)
    {
        int cur = 0;
        while (cur < data.size())
        {
            if ((data[cur] >> 3 == 0b11110) && check(data, 3, cur + 1))
            {
                cur += 4;
            }
            else if ((data[cur] >> 4 == 0b1110) && check(data, 2, cur + 1))
            {
                cur += 3;
            }
            else if ((data[cur] >> 5 == 0b110) && check(data, 1, cur + 1))
            {
                cur += 2;
            }
            else if (data[cur] >> 7 == 0b0) // Single byte char
            {
                cur += 1;
            }
            else // Not vaild UTF-8 char
            {
                return false;
            }
        }
        return true;
    }

    bool check(vector<int> &data, int size, int index) // Check vaild n-bytes character (n > 1)
    {
        while (size > 0)
        {
            size--;
            if (index >= data.size() || data[index] >> 6 != 0b10)
            {
                return false;
            }
            index++;
        }
        return true;
    }
};